### PR TITLE
Make Eclipse mostly compile gradle again

### PR DESCRIPTION
**Pull request by arming9**
Eclipse can't handle external .jars in the classpath very well. If two projects reference the same (Eclipse-external) .jar, it will index the .jar twice. Therefore changed the eclipse.gradle in a way that:

- every .jar in any project of the workspace will be represented as linked file in the gradle (root) project. 
- every project classpath is altered to reference that linked .jar as an Eclipse internal file (/gradle/.eclipsecache/the-1.jar)

Some additional changes for eclipse to work are:
- make every project use the default JRE and compiler settings of the workspace - otherwise Eclipse will complain about the @Override annotations missing or not missing at many places 
-  do some additional classpath cleanups in the project (like removing duplicate entries etc.)

### CODE REVIEW

**Comment by oehme** [comment]
Hi Armin,

first of all, hats off to making it work in Eclipse again :) This is certainly no easy task, given that all core developers use IntelliJ, so no one feels the pain if it doesn't work in Eclipse.

I would like to split this up into two parts:

1. Changes that are *needed* to make Eclipse compile again
2. Changes that are "only useful" (like the proposed performance improvement)

Would that be okay for you?

For 1), you don't need to use the default JRE and clear the jdt properties file. Instead, you can set the `eclipse.jdt.sourceCompatibility` and `eclipse.jdt.targetCompatibility` to 1.6.

For the changes in 2), I would first like to have a little better understanding what you are trying to fix. Is it really the case that JDT indexes Jars twice? That sounds pretty surprising. Can you point me to a bug report or some similar information? I'm asking because the fix you propose adds significant complexity. Since we encourage everyone to use IntelliJ for working on Gradle Core, we don't want to add a lot of complexity if it does not solve a critical problem.

Cheers,
Stefan

**Comment by arming9** [comment]
Hi Stefan,

thanks for having a look into this. We can surely split it up, as long as it still works :)

1.: Regarding the default JDK, it is possible to set that straight, but I was desperately trying to get Eclipse compile and was stuck in too many 'must/does not override' and 'JDK not found' problem messages. So putting that all back to the default JDK looked like a sound way of getting back to a situation where I can code again. If someone uses 1.7 functionality where he shouldn't, he'll face that in the build process anyway. So what do you want me to do - don't clear the properties file in line 112 and rather set the eclipse.jdt.sourceCompatibility on every project?

2.: The performance improvement is not just about comfort: If you don't have it, Eclipse won't work on commodity hardware. I think the introduction above was not clearly stated: Eclipse will re-index each .jar for each project that has this .jar in the class path over again. Having 60 or so nested projects as it is in gradle, this has the effect that you can't use Eclipse, even after waiting for a long period of time to create them. You can test it also by altering the eclipse.gradle script to 
return originalPath;
in function eclipseCacheFileAndGivePath
When doing so, you can see that the .metadata workspace directory heavily increases in size, reflecting the much larger amount of jar indexes. And you will see that Eclipse shuts down with an OutOfMemory sooner or later.

To answer your question: No, sorry, I don't have a bug ticket on Eclipse to point you at. Google does not help that much when entering terms like 'external .jar Eclipse index'. But I think this difference maybe one of the reasons why you can add .jars AND external .jars to the class path in eclipse in the first place.

And yes, IntelliJ may be the first pick to develop in gradle, but there's just this 'at the company' situation where you have a Windows machine without admin privileges but lots of virus scanners and only Eclipse. And this is what the hack is for.

**Comment by oehme** [comment]
> So what do you want me to do - don't clear the properties file in line 112 and rather set the eclipse.jdt.sourceCompatibility on every project?

Exactly, just get rid of that line and the default JRE container. The right one will be chosen based on the targetCompatibility anyway.

> Eclipse won't work on commodity hardware

It worked within a reasonable time for me when I tried last time. I'll try again and also get in touch with the JDT committers. I hope they'll have some insight on the performance problem you are describing. Can you please remove the performance-related code from this PR for now and open another one for that?

**Comment by arming9** [comment]
Hi Stefan, just coming back to tell you I have been sick last week so nothing happened until now. I'll try to get it done this week and will report back then.
Regards, Armin

**Comment by arming9** [comment]
Hi Stefan,

it is done now. 
- You were right on the jdt properties and the JRE (although keeping them leads to the strange situation that you have to have a 1.7 jre configured in eclipse which you have to name JavaSE-1.5, but I don't care - see instructions in eclipse.gradle) 
- You were right on the performance issue. Eclipse still works (but I suppose it will freeze more often compared to the performance tweak in place)

It think the current state is the minimal implementation now and you can give it another try. 

Regarding the performance tweak - how to proceed on this matter? I can show the problem and solution with a showcase multi project build. Put that on the forum, write a gist?
Would be nice if we can employ a general solution for the eclipse plugin by itself. In my workplace, we found it a neccessity, as we're having workspaces with 300 projects and each project having something like 300 .jars.

Best Regards,

Armin

**Comment by arming9** [comment]
Dear Stefan,

reporting back because I further investigated in to the Eclipse performance issue. It seems that the problem occurs only with Eclipse 4.3 (may prior as well, but I didn't check that). With Eclipse 4.4 the problem does not occur. Anyway, in case you want to check that, you can have a go with this settings.gradle:

`
// make the master build file
new File('build.gradle').text = '''
task wrapper(type: Wrapper) {
	gradleVersion = '2.12'
}

subprojects {
    apply plugin: 'java'	
    apply plugin: 'eclipse'	

	repositories {
	    jcenter()
	}
	
	dependencies {
		compile 'org.springframework:spring-web:4.2.4.RELEASE'
		compile 'org.hibernate:hibernate:3.5.4-Final'
		compile 'org.wildfly.core:wildfly-server:2.0.10.Final'
		compile 'org.apache.kafka:kafka_2.10:0.9.0.0'
		compile 'org.apache.geronimo:geronimo-system:1.0'
		compile 'org.apache.cassandra:cassandra-all:3.2.1'
		compile 'com.netflix.hystrix:hystrix-examples:1.4.23'
		compile 'net.sf.ehcache:ehcache:2.10.1'
		compile 'com.h2database:h2:1.4.191'
		compile 'org.codehaus.groovy:groovy-all:2.4.5'
		compile 'ant:ant:1.7.0'
		compile 'com.google.appengine:appengine-api-1.0-sdk:1.9.32'
		compile 'org.apache.openjpa:openjpa:2.4.0'
		compile 'org.apache.maven.plugins:maven-dependency-plugin:2.10'
		compile 'joda-time:joda-time:2.9.2'
		compile 'org.jclouds:jclouds-compute:1.6.0'
		compile 'com.netflix.archaius:archaius-core:2.0.0-rc.7'
		compile 'com.google.appengine:appengine-api-1.0-sdk:1.9.32'
		compile 'org.mockito:mockito-core:2.0.42-beta'
		compile 'org.uberfire:uberfire-io:0.7.4.Final'
		compile 'io.netty:netty-all:4.1.0.CR2'
		compile 'org.apache.tomcat:tomcat-catalina:9.0.0.M1'
		compile 'org.eclipse.jetty:jetty-server:9.3.7.v20160115'
		compile 'jboss:jboss-system:4.2.2.GA'
		compile 'org.jboss.javaee:jboss-javaee:5.0.1.GA'
	}
}
'''

for(int n = 0; n<300; n++){
	new File("subprojects/project-$n/src/main/java").mkdirs()
	include("project-$n")
}

rootProject.children.each {project ->
	project.projectDir = new File('subprojects', project.name)
}
`

but I think further investigation is pointless, as Eclipse 4.5 is out already.

Regards,

Armin

**Comment by oehme** [comment]
Hi Armin,

> It seems that the problem occurs only with Eclipse 4.3 (maybe prior releases as well, but I didn't check that). With Eclipse 4.4 the problem does not occur.

Thank you so much for digging into this issue! I agree that we don't need to fix this if it is already fixed in Eclipse itself.

> you have to have a 1.7 jre configured in eclipse which you have to name JavaSE-1.5

That should not happen. Gradle should automatically add `JavaSE-1.7` to the classpath if you specify a compatibility level of 1.7. Can you please double-check?

I have some further questions, but I'll add those as in-line comments.

Cheers,
Stefan

File: gradle/eclipse.gradle [hunk 1]
```diff
@@ -1,23 +1,82 @@
...
+// 10. Wait 5 minutes for Eclipse to calm down
+// 11. Accept that some things just don't seem to work in Eclipse, especially some of the groovy files
+
+import org.gradle.util.GUtil
+
+def subprojectsPath = new File(projectDir,'/subprojects').absolutePath
+
+apply plugin: 'eclipse'
+
+eclipse.project.file.withXml {
```
Comment (oehme) [inline gradle/eclipse.gradle]: This should not be necessary, as the root project does not have the Groovy plugin applied, so there should not be any builder. Even if there was, it should not have any source directories.

Can you make sure that you didn't have any old configuration left over? Maybe run `cleanEclipse eclipse`

File: gradle/eclipse.gradle [hunk 2]
```diff
@@ -1,23 +1,82 @@
...
+			classpath {
+				file.whenMerged { classpath ->
+					//There are classes in here not designed to be compiled, but just used in our testing
+					classpath.entries.removeAll { it.path.contains('src/integTest/resources') }
+					//Workaround for some projects referring to themselves as dependent projects
+					classpath.entries.removeAll { it.path.contains("$project.name") && it.kind=='src' }
+					classpath.entries.removeAll { it.path.contains("$project.name/build") && it.kind=='lib' }
+				}
+				file.withXml {
+					// postprocessing the xml, because I don't know how to change the sourcepath on the classpathEntry domain object
```
Comment (oehme) [inline gradle/eclipse.gradle]: I don't understand what this block does, can you elaborate?

File: gradle/eclipse.gradle [hunk 3]
```diff
@@ -1,23 +1,82 @@
...
+// Things required for this hack to work:
+//  1. Eclipse must be a recent one and needs the Groovyclipse feature. Take it from here, for example http://dist.springsource.org/snapshot/GRECLIPSE/e4.4/
+//  2. 2GB of memory is required (Add or alter -Xmx2g to eclipse.ini)
+//  3. The complete gradle project from github must be cloned (i.e. https://github.com/gradle/gradle)
+//  5. Run ./gradlew build on that folder
+//  6. Run ./gradlew eclipse on that folder
+//  7. Configure a JRE in Eclipse which is at least 1.7. Set the JRE name property of that JRE to 'JavaSE-1.5'
+//  8. In Window->Preferences->Groovy->Compiler, check 'Enable Script folder support' and Add '**/*.gradle'
+//  9. Import the gradle project into the Eclipse workspace, choose 'search for nested projects', take ALL projects
+// 10. Wait 5 minutes for Eclipse to calm down
```
Comment (oehme) [inline gradle/eclipse.gradle]: I don't think we need to tell people how waiting works ;)

File: gradle/eclipse.gradle [hunk 4]
```diff
@@ -1,23 +1,82 @@
...
+// 11. Accept that some things just don't seem to work in Eclipse, especially some of the groovy files
```
Comment (oehme) [inline gradle/eclipse.gradle]: We should merge this comment into the description in the top-level readme, basically saying something like "Building the Gradle project with Eclipse is currently limited due to Eclipse's lacking Groovy support. We recommend using IntelliJ IDEA"

File: gradle/eclipse.gradle [hunk 5]
```diff
@@ -1,23 +1,82 @@
...
+// 1. It will make most of gradle compile in eclipse
+// 2. It requires some adjustments in eclipse to work properly - see below
+//
+// Things required for this hack to work:
+//  1. Eclipse must be a recent one and needs the Groovyclipse feature. Take it from here, for example http://dist.springsource.org/snapshot/GRECLIPSE/e4.4/
+//  2. 2GB of memory is required (Add or alter -Xmx2g to eclipse.ini)
+//  3. The complete gradle project from github must be cloned (i.e. https://github.com/gradle/gradle)
+//  5. Run ./gradlew build on that folder
+//  6. Run ./gradlew eclipse on that folder
+//  7. Configure a JRE in Eclipse which is at least 1.7. Set the JRE name property of that JRE to 'JavaSE-1.5'
```
Comment (oehme) [inline gradle/eclipse.gradle]: As mentioned before, this should not be necessary.

File: gradle/eclipse.gradle [hunk 6]
```diff
@@ -1,23 +1,82 @@
...
-                }
-            }
-        }
-    }
+// This is sort of a hack
+// 1. It will make most of gradle compile in eclipse
+// 2. It requires some adjustments in eclipse to work properly - see below
+//
+// Things required for this hack to work:
+//  1. Eclipse must be a recent one and needs the Groovyclipse feature. Take it from here, for example http://dist.springsource.org/snapshot/GRECLIPSE/e4.4/
```
Comment (oehme) [inline gradle/eclipse.gradle]: Let's just point users to Eclipse 4.5 + the latest Groovy plugin instead of saying "recent".

File: gradle/eclipse.gradle [hunk 7]
```diff
@@ -1,23 +1,82 @@
...
+//  2. 2GB of memory is required (Add or alter -Xmx2g to eclipse.ini)
+//  3. The complete gradle project from github must be cloned (i.e. https://github.com/gradle/gradle)
```
Comment (oehme) [inline gradle/eclipse.gradle]: I think that is obvious, since you're looking at a Github repo.

File: gradle/eclipse.gradle [hunk 8]
```diff
@@ -1,23 +1,82 @@
...
+//  5. Run ./gradlew build on that folder
```
Comment (oehme) [inline gradle/eclipse.gradle]: Why is that necessary?

**Comment by arming9** [comment]
Hi Stefan,

I think you're asking the right questions :) 
Seems I had some zombie code left over from the Eclipse 4.3 performance adjustments. 
So I cut down the code two only 2 lines more than the original code. I hope you're not asking any more questions, otherwise no code will be left anymore and the pull request was pointless.

Regarding the JRE adjustments, I'm not sure on how to proceed.
The gradle/build.gradle file applies the gradle/gradle/groovyProject.gradle to every subproject.
In the groovyProject.gradle, you'll find 
```
sourceCompatibility = 1.5
targetCompatibility = 1.5
```
I believe this is the reason why the eclipse plugin does set the JRE to JavaSE-1.5 in every subproject.
Some of the projects indeed use JDK6 classes (maybe even JDK7, I think I have seen it, but am too lazy now to find it) classes, for example  org.gradle.model.internal.manage.schema.extract.InvalidManagedModelElementTypeException in modelCore and org.gradle.api.internal.file.collections.jdk7.Jdk7DirectoryWalker in core are using java.util.Deque (JDK1.7).

So what do we do here?
1. adjust the source and targetCompatibility in these projects?  I think the core project aims at being compatible with a 1.5 rt.jar and VM - not good.
2. set the JRE specially to 1.6/1.7 in these projects? I think this will clutter the build files and JRE 1.6/1.7 rt.jar class usage will propagate to the other subprojects from time to time, each time rendering these projects with build errors in eclipse (and nobody here for fixing it). This approach has another drawback, I find - it would have the effect that you need 2 (or even 3) JREs configured in Eclipse - which doesn't make it easier in the end. 
3. remove the JRE as done in the initial proposal and go with the standard JRE in eclipse, so nobody needs to adjust JRE names in eclipse (unless his standard jre is less than 1.6/1.7) and nobody needs to adjust JREs in the .gradle files.

Just tell me your thoughts, when we find a conclusion I'll move the setup instructions from eclipse.gradle to the README.md
 
Regards,

Armin

**Comment by oehme** [comment]
Hi Armin,

a contribution is not just measured in lines of code. You put a lot of research and deep thinking into this and as a result we learned something about Eclipse, which is great. Now we can tell others "you need at least Eclipse version 4.4 to make this work properly".

I checked and found that if you overwrite the `eclipse.jdt.sourceCompatibility`, you also need to overwrite `eclipse.jdt.javaRuntimeName` with the corresponding value (`JavaSE-1.6`). Then everything works fine.

Please go ahead and merge the instructions into the main readme and add the line I hinted above. Then this should be good to go 👍 

Thanks again,
Stefan

**Comment by arming9** [comment]
Hi Stefan,

done. Actually I won't be able to do any more changes until July, as we're going on vacation now. 
I hope it's ok now.

Regards,

Armin

**Comment by oehme** [comment]
Thank you Armin! I have merged your changes into master and did some final polishing.

It looks like there are still many compile errors because Eclipse's Groovy integration just doesn't understand some of our classes. There's not much we can do there. But at least the Java classes (which are the majority of production classes) can now be edited.

**Comment by oehme** [comment]
Code Review: https://code-review.gradle.org/cru/REVIEW-5963
